odoo/addons/base/models/assetsbundle.py

```diff
@@ -322,20 +376,75 @@ def save_attachment(self, type, content):
+    def js(self, is_minified=True):
+        extension = 'min.js' if is_minified else 'js'
+        attachments = self.get_attachments(extension, version_mode=VersionMode.INCLUDE if is_minified else VersionMode.NONE)
+
+        if attachments and (is_minified or self.last_modified < attachments[0].write_date):
+            return attachments[0]
+
+        if is_minified:
             content = ';\n'.join(asset.minify() for asset in self.javascripts)
-            return self.save_attachment('js', content)
-        return attachments[0]
+            return self.save_attachment(extension, content)
+        else:
+            return self.js_with_sourcemap()
+
+    def js_with_sourcemap(self):
+        """Create/modify the ir.attachment representing the not-minified content of the bundleJS
+        and the ir.attachment representing the linked sourcemap.
+        To avoid the loss of break points in the development tool when the bundleJS
+        has to be rebuilt, the bundleJS url must not change so we will not include
+        'self.version' in the ir.attachment url.
+        If any ir.attachment exists for the bundleJS, we create a new one (bundleJS and sourcemap),
+        else, we modify the 'raw' in the existing ir.attachments.
+
+        :return ir.attachment representing the un-minified content of the bundleJS
+        """
+        sourcemap_attachment = self.get_attachments('js.map', version_mode=VersionMode.NONE) \
+                                or self.save_attachment('js.map', '', with_version=False)
+        generator = SourceMapGenerator(
+            source_root="../../../",
+        )
+
+        content_bundle_list = []
+        content_line_count = 0
+        line_header = 6  # number of lines added by with_header()
+        for asset in self.javascripts:
+            if asset.is_transpiled:
+                # '+ 3' corresponds to the 3 lines added at the beginning of the file during transpilation.
+                generator.add_source(
+                    asset.url, asset._content, content_line_count, start_offset=line_header + 3)
+            else:
+                generator.add_source(
+                    asset.url, asset.content, content_line_count, start_offset=line_header)
+
+            content_bundle_list.append(asset.with_header(asset.content, minimal=False))
+            content_line_count += len(asset.content.split("\n")) + line_header
+
+        content_bundle = ';\n'.join(content_bundle_list) + "\n//# sourceMappingURL=" + sourcemap_attachment.url
+
+        js_attachment = self.get_attachments('js', version_mode=VersionMode.NONE)
+        if js_attachment:
+            js_attachment.write({
+                "raw": content_bundle.encode('utf8'),
+            })
+        else:
+            js_attachment = self.save_attachment('js', content_bundle, with_version=False)
+
+        generator._file = js_attachment.url
+        sourcemap_attachment.write({
+            "raw": json.dumps(generator.to_json()).encode('utf8'),
```

Comment:
Have we considered the [XSSI implications mentioned in the spec](https://sourcemaps.info/spec.html#h.h7yy76c5il9v)? There is supposed to be good tooling support for the recommended `)]}'` prefix.

That being said, I don't think this is much of a threat unless we start having user-specific contents in asset files (e.g accessible only to admins or authenticated users). In that case the contents of these assets could be exfiltrated via XSSI.

But if it's a matter of adding 4 bytes in the file, with minimal downsides, why not?

/cc @mart-e @xmo-odoo WDYT?

```suggestion
                # Store with XSSI-prevention prefix
                "raw": b")]}'" + content_bundle.encode('utf8'),
```

Reply:
> WDYT?

It's recommended by the spec and it can't hurt, so even if it's not strictly relevant right now (to our knowledge anyway) there's really no reason not to do it.